funlet-forward/functions/funlet-forward.js

```diff
@@ -0,0 +1,419 @@
+    instructions to forward the call to the given forwarding number,
+    with given caller ID and timeout, and to redirect to the given
+    fallback URL if the forwarding call fails.
+    When the caller is a restricted number, the response is modified
+    to play or say an error message instead.
+*/
+function forwardStage1(
+  response,
+  isForwardingAllowed, accessRestrictedErrorMessage, language, voice,
+  callerId, forwardingNumber, timeout, fallbackUrl
+) {
+  if ( !isForwardingAllowed ) {
+    simpleMessage(response, accessRestrictedErrorMessage, language, voice)
+    return;
+  }
+  let dialOptions = {
+    action: getForwardActionUrl( fallbackUrl ),
+  };
+  if ( callerId !== "" ) {
+    dialOptions.callerId = callerId;
+  }
+  dialOptions.timeout = timeout;
+  response.dial( dialOptions, forwardingNumber );
+}
+
+/*
+  Function: forwardStage2()
+
+  Parameters:
+    * response - Twilio.twiml.VoiceResponse, Twilio Voice response in progress
+    * isDialDone - boolean, whether the forwarding call has completed
+    * callStatus - string, the status of the forwarding call
+    * fallbackUrl - string, URL of a script with further instructions
+                    in case the forwarding call failed
+  Response:
+    Until the forwarding call has completed, the response is left unchanged.
+    When the forwarding call has ended in a failure, the response is modified
+    with instructions to redirect to the fallback URL, if any. Otherwise,
+    an instruction to hang up is added to the response.
+
+  Returns:
+    true when the forwarding call has completed,
+    false otherwise.
+*/
+function forwardStage2(response, isDialDone, callStatus, fallbackUrl) {
+  if (isDialDone) {
+    if (
+      callStatus !== "answered" &&
+      callStatus !== "completed" &&
+      fallbackUrl !== ""
+    ) {
+      response.redirect( fallbackUrl );
+    } else {
+      response.hangup();
+    }
+  }
+  return isDialDone;
+}
+
+/*
+  4. Main Handler
+
+  This is the entry point to your Twilio Function,
+  which will run to process an incoming HTTP request
+  such as the ones generated by Twilio events.
+*/
+
+exports.handler = function(context, event, callback) {
+  const NO_ERROR = null;
+
+  let
+    response = new Twilio.twiml.VoiceResponse(),
+    isDial = isDialDone(event, context, config),
+    callStatus = getCallStatus(event, context, config),
+    fallbackUrl = getFallbackUrl(event, context, config),
+    caller = getCaller(event, context, config),
+    called = getPhoneNumberCalled(event, context, config),
+    allowedCallers = getAllowedCallers(event, context, config),
+    accessRestrictedErrorMessage =
+      getAccessRestrictedErrorMessage(event, context, config),
+    language = getLanguage(event, context, config),
+    voice = getVoice(event, context, config),
+    callerId = getCallerId(event, context, config),
+    forwardingNumber = getPhoneNumber(event, context, config),
```

Comment:
The original documentation expresses strong views which are not always reflected in the actual implementation ;) The original Twimlet does not check missing parameters either, but only produces meaningful TwiML when used under the prescribed conditions.

For example, calling the Forward Twimlet without a `PhoneNumber` results in a `<Dial>` without a phone number:

```
$ curl https://twimlets.com/forward
<?xml version="1.0" encoding="UTF-8"?>
<Response><Dial action="/forward?Dial=true" timeout="20"/></Response>
```

Reply:
Hmm I feel like that's one we should fix. Because if we check for that and throw a meaningful error the user can fix it. If the user would end up with that TwiML they'll get an error that they won't understand.